### PR TITLE
JSCS Website

### PR DESCRIPTION
**Pull request by mdevils**
For a long time we carried the idea of generating JSCS documentation website using JSDoc from JSCS rules. The first attempt was made by @just-boris here: #327.

I tried to continue developing the idea:

1. To have documentation generated from rule files.
2. To have a website about JSCS.
3. To have all that generated automatically.

So here is a bunch of `gulp`-script doing all the job.

The generated example can be found here: http://jscs-dev.github.io/node-jscs/
Later I wish we move here: `jscs.info`.

@mikesherov, @markelog, @mrjoelkemp check it out, please.

### CODE REVIEW

**Comment by hzoo** [comment]
Looks awesome! Small thing but the JSCS icon links to https://jscs-dev.github.io/ which is a 404 now.

File: OVERVIEW.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
+<!-- toc -->
+
+## Presets
+
+ * [Airbnb](https://github.com/jscs-dev/node-jscs/blob/master/presets/airbnb.json) - https://github.com/airbnb/javascript
+ * [Crockford](https://github.com/jscs-dev/node-jscs/blob/master/presets/crockford.json) - http://javascript.crockford.com/code.html
+ * [Google](https://github.com/jscs-dev/node-jscs/blob/master/presets/google.json) - https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml
+ * [Grunt](https://github.com/jscs-dev/node-jscs/blob/master/presets/grunt.json) - http://gruntjs.com/contributing#syntax
+ * [jQuery](https://github.com/jscs-dev/node-jscs/blob/master/presets/jquery.json) - https://contribute.jquery.org/style-guide/js/
+ * [MDCS](https://github.com/jscs-dev/node-jscs/blob/master/presets/mdcs.json) - https://github.com/mrdoob/three.js/wiki/Mr.doob's-Code-Style%E2%84%A2
```
Comment (zxqfox) [inline OVERVIEW.md]: The link should be:
```md
[https://github.com/mrdoob/three.js/wiki/Mr.doob's-Code-Style™](https://github.com/mrdoob/three.js/wiki/Mr.doob's-Code-Style%E2%84%A2)
```

**Comment by zxqfox** [comment]
@mdevils Great work!

Do you think about regrouping rules according to @markelog's work (https://gist.github.com/markelog/752e53faf9f27de4024d)?
It should helps a lot for those who want manually fix their code style presets.

File: website/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "jscs.info",
```
Comment (zxqfox) [inline website/package.json]: `private: true` ?

Comment (mdevils) [inline website/package.json]: Thanks, will add this!

**Comment by mdevils** [comment]
> Do you think about regrouping rules according to @markelog's work (https://gist.github.com/markelog/752e53faf9f27de4024d)?

I would leave this for later work.

**Comment by mdevils** [comment]
> Looks awesome! Small thing but the JSCS icon links to https://jscs-dev.github.io/ which is a 404 now.

Thanks, will check this out!

File: OVERVIEW.md (hunk 1 above)
Comment (mdevils) [inline OVERVIEW.md]: Thanks, will fix this!

**Comment by mrjoelkemp** [comment]
This looks great. Great work! 

We need to make sure we update Contributing.md with the following:

1. Replace all instances of Readme.md with the appropriate link to the website
2. In the section "Proposing additional presets," users should modify OVERVIEW.md now, not README.md.

For the list of companies/projects that use JSCS, we can add:

* Adobe (http://www.adobe.com/)
* Grunt (http://gruntjs.com/)

**Comment by hzoo** [comment]
Definetely probably don't need that many - but you can also add:

- Ionic (http://ionicframework.com/)
- Moment.js (http://momentjs.com/)
- Ember (http://www.emberjs.com/)
- TodoMVC (http://todomvc.com/)
- Famous (http://famo.us/)
- RxJS (http://reactive-extensions.github.com/RxJS/)

**Comment by mdevils** [comment]
Thank you, guys!!

* @mrjoelkemp, @hzoo, added all the companies/products. Tell me if you have more.
* @mrjoelkemp, fixed README.md references.

File: website/package.json (hunk 2 above)
Comment (mdevils) [inline website/package.json]: Fixed

File: OVERVIEW.md (hunk 1 above)
Comment (mdevils) [inline OVERVIEW.md]: Fixed

**Comment by mdevils** [comment]
Fixed style issues, tried and published website to http://jscs.info/

**Comment by mrjoelkemp** [comment]
Looks great.

**Comment by zxqfox** [comment]
:+1: :clap:

**Comment by mdevils** [comment]
I am ready to merge this. How do you think?

**Comment by just-boris** [comment]
Hello, glad to see it!
1. Why are you are parse rules through Regexp instead of JSDoc? It may cause some issues in future when rules will become more complicated.
2. You still have a very long page with whole rules list. What about short rules index for faster navigation?

But I like general direction, and I guess it can be made later

File: OVERVIEW.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+```js
+// jscs:disable requireCurlyBraces, requireDotNotation
+if (x['a']) y(); // all errors from requireCurlyBraces OR requireDotNotation on this line will be ignored
+// jscs:enable requireCurlyBraces
+if (z['a']) a(); // all errors from requireDotNotation, but not requireCurlyBraces, will be ignored
+// jscs:enable requireDotNotation
+if (z['a']) a(); // all errors will be reported
+```
+
+## Versioning & Semver
```
Comment (zxqfox) [inline OVERVIEW.md]: I'm not sure we should move this section to overview.

Comment (mdevils) [inline OVERVIEW.md]: I think we can reconsider this later

File: README.md [hunk 4]
```diff
@@ -1,3827 +1,32 @@
...
+ * [Twitter Bootstrap](http://getbootstrap.com/)
+ * [AngularJS](https://angularjs.org/)
+ * [Yandex](https://yandex.com/)
+ * [Wikimedia](https://www.wikimedia.org/)
+ * [Grunt](http://gruntjs.com/)
+ * [Ember.js](http://www.emberjs.com/)
+ * [Ionic](http://ionicframework.com/)
+ * [Moment.js](http://momentjs.com/)
+ * [TodoMVC](http://todomvc.com/)
+ * [Famous](http://famo.us/)
```
Comment (zxqfox) [inline README.md]: License section missed?

**Comment by mdevils** [comment]
@just-boris,

> 1. Why are you are parse rules through Regexp instead of JSDoc? It may cause some issues in future when rules will become more complicated.

I believe this can be fixed later if we get issues.

> 2. You still have a very long page with whole rules list. What about short rules index for faster navigation?

I would postpone this too. First I want this to be merged to avoid complicated rebases later.

File: README.md (hunk 4 above)
Comment (mdevils) [inline README.md]: License section?

Comment (zxqfox) [inline README.md]: Yeah, I mean something like "JSCS is distributed under MIT License, see LICENSE file" etc.

Almost each repo has it.

**Comment by zxqfox** [comment]
Personally I don't like the idea to store website in the main repo. But code is lgtm. Great work!

**Comment by just-boris** [comment]
@zxqfox it is the easiest way to keep docs and code in sync

**Comment by zxqfox** [comment]
@just-boris yeah. But it should be built and published just after releasing new version so anyway it's a script that can fetch main website repo, push there autogenerated content, and publish to gh-pages or something. But without website repo you'll see a mix of website and code fixes. It's a critical difference for me ;-)

**Comment by mdevils** [comment]
@zxqfox I see you point, but it will be really difficult to keep them in sync otherwise.

**Comment by mrjoelkemp** [comment]
> First I want this to be merged to avoid complicated rebases later.

Yes. I don't see any blockers from merging. We can patch after.

**Comment by mdevils** [comment]
Merged. Thank you a lot, guys!

**Comment by hzoo** [comment]
More projects

Flight: https://flightjs.github.io/
RxJS: https://reactive-extensions.github.io/RxJS/
h5bp: http://html5boilerplate.com/
Ghost: https://ghost.org/

**Comment by mrjoelkemp** [comment]
Also FineUploader: https://github.com/FineUploader/fine-uploader

and you can fix this quickly @mdevils https://github.com/jscs-dev/node-jscs/issues/872

**Comment by mikesherov** [comment]
Great work @mdevils and @justboris !

On Friday, January 2, 2015, Joel Kemp <notifications@github.com> wrote:

> Also FineUploader: https://github.com/FineUploader/fine-uploader
>
> and you can fix this quickly @mdevils <https://github.com/mdevils> #872
> <https://github.com/jscs-dev/node-jscs/issues/872>
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/jscs-dev/node-jscs/pull/873#issuecomment-68569761>.
>


-- 
Mike Sherov

**Comment by markelog** [comment]
Oh, man, this was merged pretty fast.

1. I'd say it's definitely should be in separate repo, with it's own issues and commits, otherwise it would clutter everything
1. Readme.md of the `node-jscs` should be clean now, and consist only with very basic info about the project
1. Twitter feed, if we need one, should also show "#jscs" tweets
1. All those "who uses" references should be verified, (although i don't feel the need to publish them in the main page or publish it at all :-), i think we need a links for each of them to be sure people are actually use it.
1. "who uses" reference should be more focused, now everything is in the pile without evident structure
1. Rules should be dived and clusterized with their own pages and description for them should be taken from JSDoc as it was in @just-boris pull
1. jscs.io also could be used, as redirect
1. "Extensions", "Friendly packages", "CLI options", "options", "presets" should be moved to their own pages

I would put N1 as a priority, so everything else could be moved to the issues.

**Comment by markelog** [comment]
btw, great job!

**Comment by mikesherov** [comment]
I agree on separate repo, but glad to see we finally have something!

**Comment by alawatthe** [comment]
Nice job, I really like the website. I have just two remarks:

What do you think about having a textarea like http://jshint.com or http://jslint.com to paste code and check it online? I would like it. (I not sure if jscs is currently working in the browser.)

And to nitpick a little bit: the logo and the favicon aren't looking too good on retina displays. The logo can easily be created with SVG and for favicons I really like the http://realfavicongenerator.net/ . I'll create a pull request for this later.

**Comment by hzoo** [comment]
I think that would be cool! There is browserify setup already so that could work?

**Comment by gero3** [comment]
It works. We already use it in http://zz85.github.io/mrdoobapproves/.

**Comment by mdevils** [comment]
@markelog moved repo to https://github.com/jscs-dev/jscs-dev.github.io

> Twitter feed, if we need one, should also show "#jscs" tweets

I believe we need it, it indicates that the project is alive and in active development. About `#jscs`, try that hashtag, there is a lot of garbage, so it is useless.

> "who uses" reference should be more focused, now everything is in the pile without evident structure

I would like to have logos of the key projects on the page. That is way more friendly.

**Comment by mdevils** [comment]
> Rules should be dived and clusterized with their own pages and description for them should be taken from JSDoc as it was in @just-boris pull

Should be because...?

**Comment by mdevils** [comment]
Made an attempt on JSCS online demo: http://jscs.info/demo.html

**Comment by zxqfox** [comment]
@mdevils One more great thing we have now. Thanks! Did you think about using plugins on this page? ;-)

**Comment by mdevils** [comment]
@zxqfox Which plugins? Fell free to send PRs)

**Pull request merged**